### DISCUSSION TITLE
Create polydata hexaedron

### DISCUSSION THREAD

**Discussion by PBrockmann** [pyvista/pyvista#2503] [Q&A] [answered]
I would like to better understand how to create a set of polydata that represent hexaedrons from a climate model under construction.
Note that  by the past I have successfully created the vtk data structure to handle those output. Today I would like to use pyvista because it seems nicely embedded in jupyter notebook.

The vtk file is available from 
https://thredds-su.ipsl.fr/thredds/fileServer/ipsl_thredds/brocksce/pyvista/out_poly.vtk
and produce this with 

```
mesh_from_vtk = pyvista.PolyData('out_poly.vtk')
mesh_from_vtk.plot(show_edges=True, jupyter_backend='pythreejs', background='w')
```

![image](https://user-images.githubusercontent.com/5402758/164972072-33f4f877-fe9f-4ced-ad69-55182d3724b5.png)

So the question is more how to create correctly the polydata with pyvista ?

It is not clear for me how the different vertices create the different polydata.
Here :

```
vertices = [[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0.5, 0.5, -1]]
faces = [4, 0, 1, 2, 3, 3, 0, 1, 4, 3, 1, 2, 4]
surf = pyvista.PolyData(vertices, faces)
surf.plot(scalars=np.arange(3), show_edges=True, jupyter_backend='pythreejs', background='w')
```
 
How the 2 triangles and the square are defined ? I see only a list of vertex from the faces array.

I have then tried to create a hexeadron (6 vertex) but that crash my kernel.
![image](https://user-images.githubusercontent.com/5402758/164972546-e8911912-decd-4726-9855-a433eeb66ca5.png)

I should have 1 cell and 6 vertices.

The notebook is available from:
https://thredds-su.ipsl.fr/thredds/fileServer/ipsl_thredds/brocksce/pyvista/pyvista_02.ipynb

**Comment by akaszynski** [comment, marked as answer]
Hi @PBrockmann,

The faces in PolyData are defined by first describing the number of points in each face, followed by the indices of each face. In you example, you had:
```py
vertices = [[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0.5, 0.5, -1]]
faces = [4, 0, 1, 2, 3, 3, 0, 1, 4, 3, 1, 2, 4]
```

While this might look confusing at first, this line is actually equivalent to:
```py
face_a = [4, 0, 1, 2, 3]
face_b = [3, 0, 1, 4]
face_c = [3, 1, 2, 4]
faces = np.hstack((face_a, face_b, face_c))
```

If you look at our documentation, you can find a similar example at [pyvista.PolyData](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.html).

---

As for creating a polyhedral mesh, you'll use the same approach: Define each face with first a `6`, denoting 6 points in a face:

```py
points = <points>
face_a = [6, 0, 1, 2, 3, 4, 5]
face_b = [6, ...
faces = np.hstack((face_a, face_b, ...))

mesh = pyvista.PolyData(points, faces)
```

Of course using lists will be slower than using just `numpy`, but you get the idea and you can optimize this later if you need to.


---
### Full example

```py
import numpy as np
import pyvista

pyvista.set_plot_theme('document')

# plot three polyhedra
points = np.array(
    [[0.0480, 0.0349, 0.9982],
     [0.0305, 0.0411, 0.9987],
     [0.0207, 0.0329, 0.9992],
     [0.0218, 0.0158, 0.9996],
     [0.0377, 0.0095, 0.9992],
     [0.0485, 0.0163, 0.9987],
     [0.0572, 0.0603, 0.9965],
     [0.0390, 0.0666, 0.9970],
     [0.0289, 0.0576, 0.9979],
     [0.0582, 0.0423, 0.9974],
     [0.0661, 0.0859, 0.9941],
     [0.0476, 0.0922, 0.9946],
     [0.0372, 0.0827, 0.9959],
     [0.0674, 0.0683, 0.9954]],
)
    

face_a = [6, 0, 1, 2, 3, 4, 5]
face_b = [6, 6, 7, 8, 1, 0, 9]
face_c = [6, 10, 11, 12, 7, 6, 13]
faces = np.hstack((face_a, face_b, face_c))

mesh = pyvista.PolyData(points, faces)
mesh.plot(show_edges=True)
```
![tmp](https://user-images.githubusercontent.com/11981631/164981458-3c1e2292-ee01-4bea-a01f-bafe78cf4dc8.png)

**Reply by akaszynski** [reply]
BTW, we don't have any polydata examples. Would you mind if we could grab some of your datasets and use them as examples within [PyVista Examples](https://docs.pyvista.org/examples/index.html)?

**Reply by akaszynski** [reply]
Hello @PBrockmann, pinging you again about the example data. I'd love to demo it in our examples!

**Reply by PBrockmann** [reply]
Back from few days off.
I will answer you in details of course. Thanks

**Comment by PBrockmann** [comment]
Here is a notebook under construction.
It reads a netCDF file and variable and build the grid made of hexahedra.

![image](https://user-images.githubusercontent.com/5402758/166835386-f2feb18b-1c50-440e-8e08-69c97f9654e1.png)

https://thredds-su.ipsl.fr/thredds/fileServer/ipsl_thredds/brocksce/pyvista/pyvista_03.ipynb
